### PR TITLE
[fix] Dimension out of range in pixel_shuffle / pixel_unshuffle

### LINKED ISSUE #54051: Updating Pytorch Mobile to 1.8.0 from 1.7.1 causes “Dimension Out of Range” error in the PixelShuffle layer
Hi,

I have updated the PytorchMobile for my Android application from 1.7.1 to 1.8.0 but doing so caused an error in the inference phase. Somehow the tensor dimensions became not valid anymore for PixelShuffle layer. Here is the error:

```
    Process: com.example.pytorchtutorial, PID: 14078
    java.lang.RuntimeException: The following operation failed in the TorchScript interpreter.
    Traceback of TorchScript, serialized code (most recent call last):
      File "code/__torch__/CARN_SR/carn_x2/___torch_mangle_86.py", line 130, in forward
        _62 = getattr(self, "prepack_folding._jit_pass_packed_weight_32")
        _63 = ops.prepacked.conv2d_clamp_run(_61, _62)
        out8 = torch.pixel_shuffle(_63, 2)
               ~~~~~~~~~~~~~~~~~~~ <--- HERE
        _64 = getattr(self, "prepack_folding._jit_pass_packed_weight_33")
        out9 = ops.prepacked.conv2d_clamp_run(out8, _64)
    
    Traceback of TorchScript, original code (most recent call last):
      File "/usr/local/lib/python3.8/site-packages/torch/nn/modules/pixelshuffle.py", line 46, in forward
        def forward(self, input: Tensor) -> Tensor:
            return F.pixel_shuffle(input, self.upscale_factor)
                   ~~~~~~~~~~~~~~~ <--- HERE
    RuntimeError: Dimension out of range (expected to be in range of [-6, 5], but got 4294967290)
```

When I downgrade the Pytorch Mobile version to 1.7.1 the error disappears. I think the problem is caused in PixelShuffle layer because when I use another model for inference, the error is thrown in that layer as well. For example:

```
    Traceback of TorchScript, serialized code (most recent call last):
      File "code/__torch__/Models/ESPCN/___torch_mangle_163.py", line 29, in forward
        _10 = (_4).forward(_8, )
        _11 = (_3).forward()
        _12 = (_0).forward((_1).forward((_2).forward(_10, ), ), )
                            ~~~~~~~~~~~ <--- HERE
        _13 = torch.contiguous(_12, memory_format=0)
        return _13
      File "code/__torch__/torch/nn/modules/pixelshuffle/___torch_mangle_160.py", line 8, in forward
      def forward(self: __torch__.torch.nn.modules.pixelshuffle.___torch_mangle_160.PixelShuffle,
        argument_1: Tensor) -> Tensor:
        return torch.pixel_shuffle(argument_1, 2)
               ~~~~~~~~~~~~~~~~~~~ <--- HERE
```

I would appreciate any help with this. Thanks

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #54051

Problem was application of the unary minus operator to an unsigned type. Positive indices are now used to build the permutation array for both `pixel_shuffle` and `pixel_unshuffle`.

### CODE REVIEW

File: aten/src/ATen/native/PixelShuffle.cpp [hunk 1]
```diff
@@ -45,11 +45,11 @@ Tensor pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
...
-  // Since 2 dims were added, the correct batch dim offsets are now:
-  // -added_dims_shape.size(), ..., -7, -6.
-  std::iota(permutation.begin(), permutation.end(), -added_dims_shape.size());
-  permutation.insert(permutation.end(), {-5 /* oc */, -2 /* h */, -4 /* 1st upscale_factor */, -1 /* w */,
-                                         -3 /* 2nd upscale_factor */});
+  std::iota(permutation.begin(), permutation.end(), 0);
+  const auto NUM_BATCH_DIMS = self.dim() - NUM_NON_BATCH_DIMS;
+  permutation.insert(permutation.end(), {NUM_BATCH_DIMS /* oc */, NUM_BATCH_DIMS + 3 /* h */,
+    NUM_BATCH_DIMS + 1 /* 1st upscale factor */, NUM_BATCH_DIMS + 4 /* w */,
+    NUM_BATCH_DIMS + 2 /* 2nd upscale factor */});
```
Comment (t-vi) [inline aten/src/ATen/native/PixelShuffle.cpp:52]: I think you could also mix non-negative batch dims and negative for the last five, but this works, too.

**Review by t-vi — APPROVED** [review]
Looks good to me. I guess without 32bit CI we won't get a test here.

**Comment by t-vi** [comment]
P.S.: I'd probably keep the commit message-to-be more neutral, but I guess it's just a style thing.

File: aten/src/ATen/native/PixelShuffle.cpp (hunk 1 above)
Comment (jbschlosser) [reply aten/src/ATen/native/PixelShuffle.cpp:52]: True! I thought it'd be more straightforward to read with a consistent sign, but that's not necessarily the case.

**Pull request closed without merging**